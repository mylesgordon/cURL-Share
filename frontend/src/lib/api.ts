import { backendUrl } from './stores';
import { get } from 'svelte/store';
import type { Project, ProjectInfo } from './types';

const url = get(backendUrl);

type CreateProjectResponse = {
	id: number;
};

type ProjectAdminStatus = {
	isUserAdmin: boolean;
}

type Fetch = typeof fetch;

export async function fetchIsLoggedIn(fetch: Fetch): Promise<boolean> {
	const response = await fetch(`${url}/api/v1/user-status`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	const responseJson = await response.json();
	return responseJson.is_logged_in ? responseJson.is_logged_in : false;
}

export async function fetchProject(fetch: Fetch, projectId: string): Promise<Project> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjectAdminStatus(fetch: Fetch, projectId: string): Promise<ProjectAdminStatus> {
	const projectResponse = await fetch(`${url}/api/v1/project/${projectId}/is-user-admin`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});

	return await projectResponse.json();
}

export async function fetchProjects(fetch: Fetch): Promise<ProjectInfo[]> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'GET',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
	return await response.json();
}

export async function createProjectRequest(
	fetch: Fetch,
	projectInfo: ProjectInfo
): Promise<CreateProjectResponse> {
	const response = await fetch(`${url}/api/v1/project`, {
		method: 'POST',
		body: JSON.stringify(projectInfo),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});

	return await response.json();
}

export async function deleteProjectRequest(fetch: Fetch, project_id: number): Promise<void> {
	await fetch(`${url}/api/v1/project/${project_id}`, {
		method: 'DELETE',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}

export async function updateProjectRequest(fetch: Fetch, project: Project): Promise<void> {
	await fetch(`${url}/api/v1/project/${project.info.id}`, {
		method: 'POST',
		body: JSON.stringify(project),
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url, 'Content-Type': 'application/json' },
		credentials: 'include'
	});
}

export async function logInRequest(
	fetch: Fetch,
	endpoint: string,
	username: string,
	password: string
): Promise<number> {
	const request = await fetch(`${url}/api/v1/${endpoint}`, {
		method: 'POST',
		mode: 'cors',
		headers: {
			'Access-Control-Allow-Origin': url,
			'Content-Type': 'application/json'
		},
		body: JSON.stringify({ username, password }),
		credentials: 'include'
	});

	return request.status;
}

export async function logOutRequest(fetch: Fetch): Promise<void> {
	await fetch(`${url}/api/v1/log-out`, {
		method: 'POST',
		mode: 'cors',
		headers: { 'Access-Control-Allow-Origin': url },
		credentials: 'include'
	});
}
