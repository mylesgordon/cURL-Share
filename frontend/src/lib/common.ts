export function splitAndTrim(input: string): Array<string> {
		return input.split(',').map((item) => item.trim());
	}