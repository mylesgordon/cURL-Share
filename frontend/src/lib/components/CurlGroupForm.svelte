<script lang="ts">
	import { Button } from 'agnostic-svelte';
	import { flip } from 'svelte/animate';
	import { splitAndTrim } from '$lib/common';
	import Input from 'agnostic-svelte/components/Input/Input.svelte';

	enum ButtonType {
		Up = -1,
		Down = 1
	}
	type Test = {
		id: number;
		name: string;
		description: string;
		raw_query: string;
	};
	let testData: Array<Test> = [
		{
			id: 0,
			name: 'Test123',
			description: 'This query does something really cool - trust me.',
			raw_query: 'abc123'
		},
		{
			id: 1,
			name: 'ASD',
			description: 'This is something different. This query does something less cool.',
			raw_query: 'abc123'
		}
	];
	const flipDurationMs = 750;

	let labels: Array<string> = [];
	$: labels = splitAndTrim(labelsText);

	let description: string;
	let labelsText: string = labels.join('/');
	let name: string;
	let rawCurlInput: string;

	let maxId = Math.max(...testData.map((data) => data.id), 0);

	// cURLs
	let hovering = -1;

	$: console.log(testData);

	function addCurl() {
		maxId += 1;
		testData = [
			...testData,
			{
				id: maxId,
				name: `cURL #${maxId}`,
				description: 'Insert description here',
				raw_query: rawCurlInput
			}
		];
		rawCurlInput = '';
	}

	function onSubmit() {}

	function buttonClick(index: number, buttonType: ButtonType) {
		const itemDestination = (index + buttonType) as number;
		if (testData[itemDestination] === undefined) {
			return;
		}
		swapItems(index, itemDestination);
	}

	function swapItems(a: number, b: number) {
		[testData[a], testData[b]] = [testData[b], testData[a]];
		testData = testData;
	}

	function dragStart(event: DragEvent, itemIndex: number) {
		console.log(typeof event);
		event.dataTransfer?.setData('text/plain', itemIndex.toString());
	}

	function drop(event: DragEvent, itemIndex: number) {
		const eventString = event.dataTransfer?.getData('text/plain') ?? '';
		const eventIndex = parseInt(eventString);

		if (eventIndex === undefined) {
			return;
		}

		swapItems(eventIndex, itemIndex);
	}
</script>

<form class="flex flex-col space-y-3 p-4" on:submit|preventDefault={onSubmit}>
	<Input isRounded id="name" label="Name" bind:value={name} />
	<Input isRounded id="description" label="Description" bind:value={description} />
	<Input isRounded id="labels" label="Labels" bind:value={labelsText} />

	<span>cURLs</span>
	<!-- TODO: aria-describedby -->

	<ol class="flex flex-col space-y-2">
		{#each testData as data, index (data.id)}
			<li
				class="flex flex-col space-y-2 border rounded-md p-2"
				class:is-active={hovering === index}
				animate:flip={{ duration: flipDurationMs }}
				draggable="true"
				on:dragstart={(event) => dragStart(event, index)}
				on:dragenter|preventDefault
				on:dragover|preventDefault={() => {
					hovering = index;
				}}
				on:dragend={() => {
					hovering = -1;
				}}
				on:drop|preventDefault={(event) => {
					drop(event, index);
				}}
			>
				<Input isRounded id="name" label="Name" bind:value={data.name} />
				<Input isRounded id="description" label="Description" bind:value={data.description} />
				<Input isRounded id="raw-query" label="Raw cURL Query" bind:value={data.raw_query} />

				<div class="flex justify-end space-x-2">
					<Button
						isRounded
						on:click={() => buttonClick(index, ButtonType.Up)}
						isDisabled={index === 0}
					>
						<span class="sr-only">Move {data.name} up</span>
						<svg width="16" height="16" focusable="false" aria-hidden="true">
							<use xlink:href="#icon--up" />
						</svg>
					</Button>
					<Button
						isRounded
						on:click={() => buttonClick(index, ButtonType.Down)}
						isDisabled={index === testData.length - 1}
					>
						<span class="sr-only">Move {data.name} down</span>
						<svg width="16" height="16" focusable="false" aria-hidden="true">
							<use xlink:href="#icon--down" />
						</svg>
					</Button>
				</div>
			</li>
		{/each}
	</ol>

	<div class="flex flex-row">
		<Input isRounded id="raw-curl" label="Add cURL query" bind:value={rawCurlInput} />
		<Button isRounded on:click={addCurl} css="ml-2 self-end add-button">Add</Button>
	</div>

	<Button isBordered isRounded mode="primary" type="submit">Create</Button>
</form>

<svg xmlns="http://www.w3.org/2000/svg" class="visuallyHidden" focusable="false" aria-hidden="true">
	<defs>
		<symbol id="icon--up" viewBox="0 0 14 14">
			<path fill="none" stroke="currentColor" stroke-width="2" d="M.7 10.1L7 3.8l6.3 6.3" />
		</symbol>
		<symbol id="icon--down" viewBox="0 0 14 14">
			<path fill="none" stroke="currentColor" stroke-width="2" d="M.7 3.8L7 10.1l6.3-6.3" />
		</symbol>
	</defs>
</svg>

<style>
	.add-button {
		height: 2.375rem;
	}

	.is-active {
		background-color: rgba(143, 207, 253, 0.5);
		border: 4px dashed black;
	}
</style>
