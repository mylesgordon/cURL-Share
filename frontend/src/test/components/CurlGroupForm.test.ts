import '@testing-library/jest-dom/extend-expect';
import { createCurlGroupRequest, updateCurlGroupRequest } from '$lib/api.page';
import { goto } from '$app/navigation';
import { render, screen } from '@testing-library/svelte';
import { testMatchingSnapshot } from '../common';
import CurlGroupForm from '$lib/components/CurlGroupForm.svelte';
import userEvent from '@testing-library/user-event';
import type { ComponentProps } from 'svelte';
import type { CreateProjectResponse, Curl, CurlGroup } from '$lib/types';

vi.mock('$lib/api.page');
vi.mock('$app/navigation');

describe('CurlGroupForm component', () => {
	const testCreatingCurl: Array<Curl> = [
		{
			id: 1,
			name: 'cURL #1',
			description: 'Insert description here',
			rawQuery: 'curl -v google.com'
		}
	];

	const testEditingCurls: Array<Curl> = [
		{
			id: 1,
			name: 'cURL #1',
			description: 'Insert description here',
			rawQuery: 'curl -v google.com'
		},
		{
			id: 2,
			name: 'cURL #2',
			description: 'Insert another description here',
			rawQuery: 'curl -v google.com'
		}
	];

	const testCurlGroup: CurlGroup = {
		id: -1,
		curls: JSON.stringify(testCreatingCurl),
		description: 'Insert description',
		labels: '',
		name: 'New cURL Group',
		project_id: 0
	};

	const creatingProps: ComponentProps<CurlGroupForm> = {
		curlGroup: {
			id: -1,
			curls: '[]',
			description: 'Insert description',
			labels: '',
			name: 'New cURL Group',
			project_id: 0
		},
		editing: undefined,
		projectId: 0
	};

	const editingProps: ComponentProps<CurlGroupForm> = {
		curlGroup: {
			id: 1,
			curls: JSON.stringify(testEditingCurls),
			description: '',
			labels: '',
			name: '',
			project_id: 0
		},
		editing: true,
		projectId: 0
	};

	testMatchingSnapshot(CurlGroupForm, creatingProps);
	testMatchingSnapshot(CurlGroupForm, editingProps);

	it('calls the correct endpoint for creating a cURL group', async () => {
		render(CurlGroupForm, creatingProps);

		const response: CreateProjectResponse = { id: 1 };
		createCurlGroupRequest.mockResolvedValue(response);

		const curlInput = await screen.findByLabelText('Add cURL query');
		const addButton = await screen.findByText('Add');
		const createButton = await screen.findByText('Create');

		await userEvent.type(curlInput, 'curl -v google.com');
		await userEvent.click(addButton);
		await userEvent.click(createButton);

		expect(createCurlGroupRequest).toHaveBeenCalledWith(window.fetch, testCurlGroup);
		expect(goto).toHaveBeenCalledWith('/project/0/group/1');
	});

	it('calls the correct endpoint for editing a cURL group', async () => {
		render(CurlGroupForm, editingProps);

		updateCurlGroupRequest.mockResolvedValue();

		const finishButton = await screen.findByText('Finish');
		await userEvent.click(finishButton);

		expect(updateCurlGroupRequest).toHaveBeenCalledWith(window.fetch, editingProps.curlGroup);
		expect(goto).toHaveBeenCalledWith('/project/0/group/1');
	});
});
