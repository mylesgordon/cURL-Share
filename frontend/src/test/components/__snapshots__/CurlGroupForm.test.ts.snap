// Vitest Snapshot v1

exports[`CurlGroupForm component > should match snapshot 1`] = `
<body>
  <div>
    <form
      class="flex flex-col space-y-3 p-4 s-07XP8E40ha51"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="name"
        >
          Name
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="name"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="description"
        >
          Description
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="description"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="labels"
        >
          Labels (comma separated)
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="labels"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <span
        class="s-07XP8E40ha51"
      >
        cURLs
      </span>
       
      <ol
        class="flex flex-col space-y-2 s-07XP8E40ha51"
      />
       
      <div
        class="flex flex-row s-07XP8E40ha51"
      >
        <div
          class="w-100 s-JZthjkXlZrq3"
        >
          <label
            class="label s-JZthjkXlZrq3"
            for="raw-curl"
          >
            Add cURL query
          </label>
           
          <input
            class="input input-rounded s-JZthjkXlZrq3"
            id="raw-curl"
            type="text"
          />
           
        </div>
        <!--&lt;Input&gt;-->
         
        <button
          class="btn btn-rounded ml-2 self-end add-button s-NAQuwh5ehqbV"
          type="button"
        >
          Add
        </button>
        
        <!--&lt;Button&gt;-->
      </div>
       
      <button
        class="btn btn-primary btn-bordered btn-rounded s-NAQuwh5ehqbV"
        type="submit"
      >
        Create
        
      </button>
      
      <!--&lt;Button&gt;-->
    </form>
     
    <svg
      aria-hidden="true"
      class="visuallyHidden s-07XP8E40ha51"
      focusable="false"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs
        class="s-07XP8E40ha51"
      >
        <symbol
          class="s-07XP8E40ha51"
          id="icon--up"
          viewBox="0 0 14 14"
        >
          <path
            class="s-07XP8E40ha51"
            d="M.7 10.1L7 3.8l6.3 6.3"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          />
        </symbol>
        <symbol
          class="s-07XP8E40ha51"
          id="icon--down"
          viewBox="0 0 14 14"
        >
          <path
            class="s-07XP8E40ha51"
            d="M.7 3.8L7 10.1l6.3-6.3"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          />
        </symbol>
      </defs>
    </svg>
    <!--&lt;CurlGroupForm&gt;-->
  </div>
</body>
`;

exports[`CurlGroupForm component > should match snapshot 2`] = `
<body>
  <div>
    <form
      class="flex flex-col space-y-3 p-4 s-07XP8E40ha51"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="name"
        >
          Name
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="name"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="description"
        >
          Description
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="description"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label s-JZthjkXlZrq3"
          for="labels"
        >
          Labels (comma separated)
        </label>
         
        <input
          class="input input-rounded s-JZthjkXlZrq3"
          id="labels"
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <span
        class="s-07XP8E40ha51"
      >
        cURLs
      </span>
       
      <ol
        class="flex flex-col space-y-2 s-07XP8E40ha51"
      >
        <li
          class="flex flex-col space-y-2 border rounded-md p-2 s-07XP8E40ha51"
        >
          <div
            class="flex flex-row s-07XP8E40ha51"
          >
            <div
              aria-hidden="true"
              class="flex items-center pr-2 cursor-grab s-07XP8E40ha51"
              draggable="true"
            >
              <svg
                height="1.2em"
                viewBox="0 0 1792 1792"
                width="1.2em"
              >
                <path
                  d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19t-19-45v-128h-384v384h128q26 0 45 19t19 45t-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45t45-19h128v-384H384v128q0 26-19 45t-45 19t-45-19L19 941Q0 922 0 896t19-45l256-256q19-19 45-19t45 19t19 45v128h384V384H640q-26 0-45-19t-19-45t19-45L851 19q19-19 45-19t45 19l256 256q19 19 19 45t-19 45t-45 19h-128v384h384V640q0-26 19-45t45-19t45 19l256 256q19 19 19 45z"
                  fill="currentColor"
                />
              </svg>
              <!--&lt;Arrows&gt;-->
            </div>
             
            <div
              class="pt-2 pb-2 pr-2 w-full flex flex-col space-y-2 s-07XP8E40ha51"
            >
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="name"
                >
                  Name:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="name"
                  label="Name:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="description"
                >
                  Description:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="description"
                  label="Description:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="raw-query"
                >
                  Raw cURL Query:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="raw-query"
                  label="Raw cURL Query:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex justify-end space-x-2 s-07XP8E40ha51"
              >
                <button
                  class="btn btn-rounded disabled s-NAQuwh5ehqbV"
                  disabled=""
                  type="button"
                >
                  <span
                    class="sr-only s-07XP8E40ha51"
                  >
                    Move 
                    cURL #1
                     up
                  </span>
                   
                  <svg
                    aria-hidden="true"
                    class="s-07XP8E40ha51"
                    focusable="false"
                    height="16"
                    width="16"
                  >
                    <use
                      class="s-07XP8E40ha51"
                      xlink:href="#icon--up"
                    />
                  </svg>
                </button>
                
                <!--&lt;Button&gt;-->
                 
                <button
                  class="btn btn-rounded s-NAQuwh5ehqbV"
                  type="button"
                >
                  <span
                    class="sr-only s-07XP8E40ha51"
                  >
                    Move 
                    cURL #1
                     down
                  </span>
                   
                  <svg
                    aria-hidden="true"
                    class="s-07XP8E40ha51"
                    focusable="false"
                    height="16"
                    width="16"
                  >
                    <use
                      class="s-07XP8E40ha51"
                      xlink:href="#icon--down"
                    />
                  </svg>
                </button>
                
                <!--&lt;Button&gt;-->
              </div>
            </div>
          </div>
           
        </li>
        <li
          class="flex flex-col space-y-2 border rounded-md p-2 s-07XP8E40ha51"
        >
          <div
            class="flex flex-row s-07XP8E40ha51"
          >
            <div
              aria-hidden="true"
              class="flex items-center pr-2 cursor-grab s-07XP8E40ha51"
              draggable="true"
            >
              <svg
                height="1.2em"
                viewBox="0 0 1792 1792"
                width="1.2em"
              >
                <path
                  d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19t-19-45v-128h-384v384h128q26 0 45 19t19 45t-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45t45-19h128v-384H384v128q0 26-19 45t-45 19t-45-19L19 941Q0 922 0 896t19-45l256-256q19-19 45-19t45 19t19 45v128h384V384H640q-26 0-45-19t-19-45t19-45L851 19q19-19 45-19t45 19l256 256q19 19 19 45t-19 45t-45 19h-128v384h384V640q0-26 19-45t45-19t45 19l256 256q19 19 19 45z"
                  fill="currentColor"
                />
              </svg>
              <!--&lt;Arrows&gt;-->
            </div>
             
            <div
              class="pt-2 pb-2 pr-2 w-full flex flex-col space-y-2 s-07XP8E40ha51"
            >
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="name"
                >
                  Name:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="name"
                  label="Name:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="description"
                >
                  Description:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="description"
                  label="Description:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex flex-col flex-grow overflow-hidden s-Et3LeF16QMsX"
              >
                <label
                  class="s-Et3LeF16QMsX"
                  for="raw-query"
                >
                  Raw cURL Query:
                </label>
                 
                <input
                  class="input s-Et3LeF16QMsX"
                  id="raw-query"
                  label="Raw cURL Query:"
                />
              </div>
              <!--&lt;AgnosticInput&gt;-->
               
              <div
                class="flex justify-end space-x-2 s-07XP8E40ha51"
              >
                <button
                  class="btn btn-rounded s-NAQuwh5ehqbV"
                  type="button"
                >
                  <span
                    class="sr-only s-07XP8E40ha51"
                  >
                    Move 
                    cURL #2
                     up
                  </span>
                   
                  <svg
                    aria-hidden="true"
                    class="s-07XP8E40ha51"
                    focusable="false"
                    height="16"
                    width="16"
                  >
                    <use
                      class="s-07XP8E40ha51"
                      xlink:href="#icon--up"
                    />
                  </svg>
                </button>
                
                <!--&lt;Button&gt;-->
                 
                <button
                  class="btn btn-rounded disabled s-NAQuwh5ehqbV"
                  disabled=""
                  type="button"
                >
                  <span
                    class="sr-only s-07XP8E40ha51"
                  >
                    Move 
                    cURL #2
                     down
                  </span>
                   
                  <svg
                    aria-hidden="true"
                    class="s-07XP8E40ha51"
                    focusable="false"
                    height="16"
                    width="16"
                  >
                    <use
                      class="s-07XP8E40ha51"
                      xlink:href="#icon--down"
                    />
                  </svg>
                </button>
                
                <!--&lt;Button&gt;-->
              </div>
            </div>
          </div>
           
        </li>
      </ol>
       
      <div
        class="flex flex-row s-07XP8E40ha51"
      >
        <div
          class="w-100 s-JZthjkXlZrq3"
        >
          <label
            class="label s-JZthjkXlZrq3"
            for="raw-curl"
          >
            Add cURL query
          </label>
           
          <input
            class="input input-rounded s-JZthjkXlZrq3"
            id="raw-curl"
            type="text"
          />
           
        </div>
        <!--&lt;Input&gt;-->
         
        <button
          class="btn btn-rounded ml-2 self-end add-button s-NAQuwh5ehqbV"
          type="button"
        >
          Add
        </button>
        
        <!--&lt;Button&gt;-->
      </div>
       
      <button
        class="btn btn-primary btn-bordered btn-rounded s-NAQuwh5ehqbV"
        type="submit"
      >
        Finish
        
      </button>
      
      <!--&lt;Button&gt;-->
    </form>
     
    <svg
      aria-hidden="true"
      class="visuallyHidden s-07XP8E40ha51"
      focusable="false"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs
        class="s-07XP8E40ha51"
      >
        <symbol
          class="s-07XP8E40ha51"
          id="icon--up"
          viewBox="0 0 14 14"
        >
          <path
            class="s-07XP8E40ha51"
            d="M.7 10.1L7 3.8l6.3 6.3"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          />
        </symbol>
        <symbol
          class="s-07XP8E40ha51"
          id="icon--down"
          viewBox="0 0 14 14"
        >
          <path
            class="s-07XP8E40ha51"
            d="M.7 3.8L7 10.1l6.3-6.3"
            fill="none"
            stroke="currentColor"
            stroke-width="2"
          />
        </symbol>
      </defs>
    </svg>
    <!--&lt;CurlGroupForm&gt;-->
  </div>
</body>
`;
